staker/staker.go

```diff
@@ -417,11 +362,85 @@ func (s *Staker) Initialize(ctx context.Context) error {
 
 		return s.blockValidator.InitAssumeValid(stakedInfo.AfterState().GlobalState)
 	}
+	return s.setupFastConfirmation(ctx)
+}
+
+// setupFastConfirmation sets the enableFastConfirmation and fastConfirmSafe variables of staker
+// based on the config, the wallet address, and the on-chain rollup designated fast confirmer.
+// Before this function, both variables should be their default (i.e. fast confirmation is disabled).
+func (s *Staker) setupFastConfirmation(ctx context.Context) error {
+	if !s.config.EnableFastConfirmation || s.wallet.Address() == nil {
+		return nil
+	}
+	walletAddress := *s.wallet.Address()
+	client := s.l1Reader.Client()
+	rollup, err := rollupgen.NewRollupUserLogic(s.rollupAddress, client)
+	if err != nil {
+		return err
+	}
+	callOpts := s.getCallOpts(ctx)
+	fastConfirmer, err := rollup.AnyTrustFastConfirmer(callOpts)
+	if err != nil {
+		if headerreader.ExecutionRevertedRegexp.MatchString(err.Error()) {
+			log.Debug("Rollup contract does not support fast confirmation", "rollup", s.rollupAddress, "err", err)
+			return nil
+		}
+		return fmt.Errorf("getting rollup fast confirmer address: %w", err)
+	}
+	if fastConfirmer == walletAddress {
+		// We can directly fast confirm nodes
+		s.enableFastConfirmation = true
+		return nil
+	} else if fastConfirmer == (common.Address{}) {
+		// No fast confirmer enabled
+		return nil
+	}
+	// Only use gnosis safe fast confirmation, if the safe address is different from the wallet address, else it's not a safe contract.
+	codeAt, err := client.CodeAt(context.Background(), fastConfirmer, nil)
+	if err != nil {
+		return fmt.Errorf("getting code at fast confirmer address: %w", err)
+	}
+	if len(codeAt) == 0 {
+		// The fast confirmer address is an EOA address, but it does not match the wallet address so cannot enable fast confirmation.
+		log.Info("Fast confirmer address is an EOA address which does not match the wallet address so cannot enable fast confirmation", "fastConfirmer", fastConfirmer, "wallet", walletAddress)
+		return nil
+	}
+	// The fast confirmer address is a contract address, not sure if it's a safe contract yet.
+	fastConfirmSafe, err := NewFastConfirmSafe(
+		callOpts,
+		fastConfirmer,
+		s.builder,
+		s.wallet,
+		s.config.gasRefunder,
+		s.l1Reader,
+	)
+	if err != nil {
+		if headerreader.ExecutionRevertedRegexp.MatchString(err.Error()) {
+			// If the safe is not a safe contract, we can't use it for fast confirmation
+			log.Warn("Fast confirmer address is not a safe contract so cannot enable fast confirmation", "fastConfirmer", fastConfirmer, "wallet", walletAddress, "err", err)
+			return nil
+		}
+		// Unknown while loading the safe contract.
+		return fmt.Errorf("loading fast confirm safe: %w", err)
+	}
+	// Fast confirmer address is a safe contract.
```

Comment:
nit about the comment: we're not really sure it's a Safe contract, just that it implements some relevant parts of the safe API,